# Practice Case - pymongo (MongoDB)
### By Firdaus Adi Nugroho (Ido)
----------------------------------------------------------------------------------------------------------------------------

### Case
1. Membuat collection baru dengan nama clean_movies yang sama persis dengan movies collection pada database sample_mflix dengan memakai collection movies_intial (projecting_queries pymongo)
2. Validasi collection yang telah dibuat dengan parameter sbb :
    - Semua document pada clean_movies dan movie sama
    - Banyak document pada clean_movies dan movie sama
    - Semua fields pada clean_movies ada pada movie
    - Semua value pada clean_movies sama dengan semua value pada movies dengan urutan yang sama


# 1.  Import Library

In [1]:
import pymongo
from pymongo import MongoClient
import pprint
from IPython.display import clear_output
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings("ignore", category=DeprecationWarning)

# 2. Melakukan Connection Database dari Cluster

## 2.1 Movies_Initial Collection

In [2]:
cl_initial = MongoClient('mongodb+srv://adminuser12345:adminuser12345@cluster0-phmf4.gcp.mongodb.net/test?retryWrites=true&w=majority')

Melakukan Connection Cluster dari Cluster MongoDB yang dibuat dari MongoDB Atlas atau MongoDB Compass

In [3]:
#show all database in cluster
pprint.pprint(cl_initial.list_database_names())

['movies_initial', 'sample_mflix', 'admin', 'local']


Menampilkan tampilan database yang ada pada Cluster MongoDB

In [4]:
dbs = cl_initial['sample_mflix']
col = dbs.movies_initial

pprint.pprint(dbs.list_collection_names())

['clean_movies_damar',
 'clean_movies_zumar',
 'clean_movies_andreas',
 'clean_movies_febi',
 'clean_movies_indra_1',
 'clean_movies_frans',
 'clean_movies_firdaus',
 'clean_movies_tunggal',
 'movies_initial',
 'clean_movies_firli']


Menampilkan tampilan collection(table) yang ada pada database sample_mflix<br>
Selain itu mendefine tabel agar mudah untuk memudahkan dalam set and get data

In [5]:
dbs.movies_initial.find_one()

{'_id': ObjectId('5db0085107645f129de5458a'),
 'awards': '',
 'cast': 'Carmencita',
 'country': 'USA',
 'director': 'William K.L. Dickson',
 'fullplot': 'Performing on what looks like a small wooden stage, wearing a dress with a hoop skirt and white high-heeled pumps, Carmencita does a dance with kicks and twirls, a smile always on her face.',
 'genre': 'Documentary, Short',
 'imdbID': 1,
 'imdbRating': 5.9,
 'imdbVotes': 1032,
 'language': '',
 'lastupdated': '2015-08-26 00:03:45.040000000',
 'metacritic': '',
 'plot': 'Performing on what looks like a small wooden stage, wearing a dress with a hoop skirt and white high-heeled pumps, Carmencita does a dance with kicks and twirls, a smile always on her face.',
 'poster': 'http://ia.media-imdb.com/images/M/MV5BMjAzNDEwMzk3OV5BMl5BanBnXkFtZTcwOTk4OTM5Ng@@._V1_SX300.jpg',
 'rating': 'NOT RATED',
 'released': '',
 'runtime': '1 min',
 'title': 'Carmencita',
 'type': 'movie',
 'writer': '',
 'year': 1894}

Pada collection **movies_initial** terdapat beberapa kolom yaitu: _id, awards, cast, country, director, fullplot, genre, imdbID, imdbRating, imdbVotes, language, lastupdated, metacritic, plot, poster, rating, released, runtime, title, type, writer, dan year<br>
nantinya isi dari collection **movies_initial** akan **disamakan** dengan isi **collection movies** dari **field** dan **document**nya.

In [6]:
pd.DataFrame(list(col.find().limit(5)))

_id  awards                      cast country  \
0  5db0085107645f129de5458a                        Carmencita     USA   
1  5db0085107645f129de5458b  1 win.  Charles Kayser, John Ott     USA   
2  5db0085107645f129de5458c                                    France   
3  5db0085107645f129de5458d                          Fred Ott     USA   
4  5db0085107645f129de5458e                                    France   

               director                                           fullplot  \
0  William K.L. Dickson  Performing on what looks like a small wooden s...   
1  William K.L. Dickson  A stationary camera looks at a large anvil wit...   
2         �mile Reynaud  One night, Arlequin come to see his lover Colo...   
3  William K.L. Dickson  A man (Edison's assistant) takes a pinch of sn...   
4         Louis Lumi�re  A man opens the big gates to the Lumi�re facto...   

                      genre  imdbID  imdbRating  imdbVotes  ...  metacritic  \
0        Documentary, Short       1         5.9       1032  ...               
1                     Short       5         6.2       1189  ...               
2  Animation, Comedy, Short       3         6.7        566  ...               
3        Documentary, Short       8         5.9        988  ...               
4        Documentary, Short      10         6.9       3469  ...               

                                                plot  \
0  Performing on what looks like a small wooden s...   
1  Three men hammer on an anvil and pass a bottle...   
2  One night, Arlequin come to see his lover Colo...   
3  A man (Thomas Edison's assistant) takes a pinc...   
4  A man opens the big gates to the Lumi�re facto...   

                                              poster     rating    released  \
0  http://ia.media-imdb.com/images/M/MV5BMjAzNDEw...  NOT RATED               
1                                                       UNRATED  1893-05-09   
2                                                                1892-10-28   
3                                                                1894-01-09   
4                                                                1895-03-22   

  runtime                                   title   type writer  year  
0   1 min                              Carmencita  movie         1894  
1   1 min                        Blacksmith Scene  movie         1893  
2   4 min                          Pauvre Pierrot  movie         1892  
3   1 min  Edison Kinetoscopic Record of a Sneeze  movie         1894  
4   1 min   Employees Leaving the Lumi�re Factory  movie         1895  

[5 rows x 22 columns]

Berikut merupakan tampilan DataFrame dari collection movies_initial

## 2.2 Movies Collection

In [7]:
clients_movies = MongoClient('mongodb+srv://userstudent:admin1234@cluster0-nnbxe.gcp.mongodb.net/test?retryWrites=true&w=majority')

Melakukan Connection Cluster dari Cluster MongoDB yang dibuat dari MongoDB Atlas atau MongoDB Compass. Mengapa perlu berbeda cluster. Hal ini dilakukan karena pada cluster ini terlalu penuh untuk membuat collection baru.

In [8]:
#show all database in cluster
pprint.pprint(clients_movies.list_database_names())

['sample_mflix', 'admin', 'local']


Menampilkan tampilan database yang ada pada Cluster MongoDB

In [9]:
db_movies = clients_movies['sample_mflix']
moviess = db_movies.movies
# show all collecton in used database
pprint.pprint(db_movies.list_collection_names())

['clean_movies_dwilaras',
 'clean_movies_nurlailiis',
 'clean_movies_devita',
 'movies_initial',
 'clean_movies_frans',
 'clean_movies_tara',
 'clean_movies_tunggal',
 'clean_movies',
 'users',
 'clean_movies_nurrimah',
 'movies_testtopik',
 'movies_scratch',
 'clean_movies_topik',
 'movies',
 'clean_movies_afifah',
 'clean_movies_andreas',
 'clean_movies_indra_1']


Menampilkan tampilan collection(table) yang ada pada database sample_mflix<br>
Selain itu mendefine tabel agar mudah untuk memudahkan dalam set and get data

In [10]:
db_movies.movies.find_one()

{'_id': ObjectId('573a1390f29313caabcd5a93'),
 'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
 'cast': ['Howard C. Hickman', 'Enid Markey', 'Lola May', 'Kate Bruce'],
 'countries': ['USA'],
 'directors': ['Reginald Barker',
  'Thomas H. Ince',
  'Raymond B. West',
  'Walter Edwards',
  'David Hartford',
  'Jay Hunt',
  'J. Parker Read Jr.'],
 'fullplot': 'Allegorical film about peace. A king starts a war, many of the women are against it, people are pressed into service. A count has constructed a submarine and gets the order to sink an ocean liner, that is also carrying - supposedly - ammunition for the enemy. The count refuses to fire the torpedos, and sinks the submarine. He survives, but in a limbo between death and life where he meets Jesus, who takes him over to preach peace. Naturally the king arrests him and sentences him to death for treason, but then Jesus shows him the real face of war.',
 'genres': ['Drama'],
 'imdb': {'id': 6517, 'rating': 6.3, 'votes': 162},
 '

Pada collection **movies** terdapat beberapa kolom yaitu: _id, awards, cast, countries, directors, fullplot, genres, imdb, lasupdated, num_flix_comment, plot, poster, released, runtime, title, tomatoes, viewer, type, writers, year<br>
Collection **movies** merupakan **source data** untuk pembuatan collection baru yang dinamanakan **clean_movies_firdaus**

In [11]:
pd.DataFrame(list(moviess.find().limit(5)))

_id  \
0  573a1390f29313caabcd5a93   
1  573a1390f29313caabcd60e4   
2  573a1391f29313caabcd71f5   
3  573a1391f29313caabcd73f4   
4  573a1391f29313caabcd7da6   

                                              awards  \
0    {'wins': 1, 'nominations': 0, 'text': '1 win.'}   
1    {'wins': 1, 'nominations': 0, 'text': '1 win.'}   
2  {'wins': 0, 'nominations': 1, 'text': '1 nomin...   
3    {'wins': 1, 'nominations': 0, 'text': '1 win.'}   
4    {'wins': 1, 'nominations': 0, 'text': '1 win.'}   

                                                cast countries  \
0  [Howard C. Hickman, Enid Markey, Lola May, Kat...     [USA]   
1  [Charles Chaplin, Edna Purviance, Eric Campbel...     [USA]   
2     [Harold Lloyd, Mildred Davis, Anna Mae Bilson]     [USA]   
3                                    [Buster Keaton]     [USA]   
4  [Frank Mayo, Virginia Valli, Ford Sterling, Ni...     [USA]   

                                           directors  \
0  [Reginald Barker, Thomas H. Ince, Raymond B. W...   
1                                  [Charles Chaplin]   
2                      [Fred C. Newmeyer, Hal Roach]   
3                   [Edward F. Cline, Buster Keaton]   
4                                       [King Vidor]   

                                            fullplot                   genres  \
0  Allegorical film about peace. A king starts a ...                  [Drama]   
1  Charlie is on his way to the USA. He wins in a...   [Short, Comedy, Drama]   
2  Mary is looking after a young child whose pare...          [Comedy, Short]   
3  Through a series of mistaken identities Buster...  [Short, Comedy, Family]   
4  Millie Stope lives with her grandfather on a r...         [Drama, Romance]   

                                          imdb  languages  \
0    {'rating': 6.3, 'votes': 162, 'id': 6517}        NaN   
1   {'rating': 7.8, 'votes': 4680, 'id': 8133}  [English]   
2   {'rating': 6.8, 'votes': 489, 'id': 12512}  [English]   
3  {'rating': 7.8, 'votes': 3738, 'id': 13025}  [English]   
4   {'rating': 7.1, 'votes': 327, 'id': 15498}  [English]   

                     lastupdated  ...   \
0  2015-04-30 02:13:27.863000000  ...    
1  2015-09-17 04:52:02.293000000  ...    
2  2015-06-25 00:02:16.613000000  ...    
3  2015-09-16 06:21:19.243000000  ...    
4  2015-08-13 00:31:55.567000000  ...    

                                                plot  \
0  Christ takes on the form of a pacifist count t...   
1  Charlie is an immigrant who endures a challeng...   
2  A young man, unaccustomed to children, must ac...   
3  A series of mishaps manages to make a young ma...   
4  Millie Stope lives with her grandfather on a r...   

                                              poster    rated   released  \
0  https://m.media-amazon.com/images/M/MV5BMjAwNT...      NaN 1916-06-02   
1  https://m.media-amazon.com/images/M/MV5BMTNkYW...  UNRATED 1917-06-17   
2  https://m.media-amazon.com/images/M/MV5BYjgzYz...   PASSED 1921-03-27   
3                                                NaN  UNRATED 1922-03-01   
4  https://m.media-amazon.com/images/M/MV5BMjA4OT...   PASSED 1924-01-20   

  runtime          title                                           tomatoes  \
0      78   Civilization  {'viewer': {'rating': 0.0, 'numReviews': 7}, '...   
1      30  The Immigrant  {'viewer': {'rating': 4.1, 'numReviews': 636, ...   
2      35   Now or Never  {'viewer': {'rating': 3.8, 'numReviews': 3}, '...   
3      22           Cops  {'viewer': {'rating': 4.0, 'numReviews': 498, ...   
4      88   Wild Oranges  {'viewer': {'rating': 4.2, 'numReviews': 1519,...   

    type                                            writers  year  
0  movie                              [C. Gardner Sullivan]  1916  
1  movie                                                NaN  1917  
2  movie      [H.M. Walker (titles), Sam Taylor (scenario)]  1921  
3  movie                   [Buster Keaton, Edward F. Cline]  1922  
4  movie  [Joseph Hergesheimer (by), King Vidor (a

Berikut merupakan isi tabel dari **movies**. Dari perbedaan data **movies_initial** dengan **movies** terdapat perbedaan kolom dan isi dari setiap kolomnya

# 3. Eksisting File

In [12]:
#jumlah document pada table movies_initial
dbs.movies_initial.count()

46014

In [13]:
#jumlah document pada table movies
db_movies.movies.count()

23541

Terlihat bahwa **perbedaan document** antara dua collection antara **movies_initial** sebanyak **46014** dan **collection movies** sebanyak **23541**. Untuk itu pada collection **clean_movies_firdaus** merupakan hasil dari **projection** dari movies initial yang data source nya berasalh dari collection movies

## 3.1 Perbedaan Data

In [14]:
#movies_initial collection
list(col.find({},{'_id':0,'country':1,'writer':1}).limit(5))

[{'country': 'USA', 'writer': ''},
 {'country': 'USA', 'writer': ''},
 {'country': 'France', 'writer': ''},
 {'country': 'USA', 'writer': ''},
 {'country': 'France', 'writer': ''}]

In [15]:
#movies collection
list(moviess.find({},{'_id':0,'countries':1,'writers':1}).limit(5))

[{'countries': ['USA'], 'writers': ['C. Gardner Sullivan']},
 {'countries': ['USA']},
 {'countries': ['USA'],
  'writers': ['H.M. Walker (titles)', 'Sam Taylor (scenario)']},
 {'countries': ['USA'], 'writers': ['Buster Keaton', 'Edward F. Cline']},
 {'countries': ['USA'],
  'writers': ['Joseph Hergesheimer (by)',
   'King Vidor (adapted by)',
   'Tom Miranda (titles)']}]

Kita bisa melihat contoh perbedaan dari setiap isi dari data setiap collection sehingga butuh **penyesuian** (Perlu di projection)<br>
**Perbedaan:**<br>
Pada perbedaan ini hanya sebagian dari setiap fieldnya. contohnya pada **country** dan **writer**
<br>Pada collection movies_intial bernama **country dan writer** sedangkan movies collection bernama **countries dan writers**, movie_initial berbentuk **string** biasa dan movies berbentuk **list**. Sehingga perlu di**rename** dan di**split**

## 4. Menentukan Key Equalizer

In [16]:
imdb = list(moviess.find({},{'_id':0,'imdb':1}))

imdb_list = []
for i in imdb:
    imdb_list.append(list(i.values())[0])
imdb_list[:5]

[{'id': 6517, 'rating': 6.3, 'votes': 162},
 {'id': 8133, 'rating': 7.8, 'votes': 4680},
 {'id': 12512, 'rating': 6.8, 'votes': 489},
 {'id': 13025, 'rating': 7.8, 'votes': 3738},
 {'id': 15498, 'rating': 7.1, 'votes': 327}]

Hasil diatas dilakukan agar movies_init, dan movies bisa **seragam urutannya**. Hal itu dilakukan untuk projection agar value nya bisa **seragam**

# 5.  Membuat Collection Baru dari movies_initial dari movies Menggunakan Projection

In [33]:
pipeline = [
    {
        '$project': {
            ### renaming the keys ###
            '_id':1,
            
             'awards': {'nominations': {'$arrayElemAt':[{'$split':[{'$arrayElemAt':[{'$split': [ "$awards", " nomination" ]}, -2]},' ']},-1]},
                       'text': "$awards",
                       'wins': {'$arrayElemAt':[{'$split':[{'$arrayElemAt':[{'$split': [ "$awards", " win" ]}, -2]},' ']},-1]}
                      },
            
            'cast':{"$cond": [{"$eq": ["$cast", ""]},
                                    "$nonExistinField",
                                    "$cast"]},
            
            'countries': {"$cond": [{"$eq": ["$country", ""]},
                                    
                                              "$nonExistinField",
                                              {'$split': ["$country", ", "]},]},
            
            'directors': {"$cond": [{"$eq": ["$director", ""]},
                                              "$nonExistinField",
                                              {'$split': ["$director", ", "]},]},
            
            'fullPlot':{"$cond": [{"$eq": ["$fullplot", ""]},
                                    "$nonExistinField",
                                    "$fullplot"]},
            
            'genres': {"$cond": [{"$eq": ["$genre", ""]},
                                            "$nonExistinField",
                                            {'$split': ["$genre", ", "]},]},
            'imdb': {
                'id': "$imdbID",
                'rating': '$imdbRating',
                'votes': '$imdbVotes'
                },
            'languages':  {"$cond": [{"$eq": ["$language", ""]},
                                              "$nonExistinField",
                                              {'$split': ["$language", ", "]},]},
            
            'lastupdated':{"$cond": [{"$eq": ["$lastupdated", ""]},
                                    "$nonExistinField",
                                    "$lastupdated"]},
            
            'metacritic':{"$cond": [{"$eq": ["$metacritic", ""]},
                                    "$nonExistinField",
                                    "$metacritic"]},
            
            'num_mflix_comments':  {"$cond": [{"$eq": ["$num_mflix_comments", ""]},
                                              "$nonExistinField",
                                              {'$split': ["$num_mflix_comments", ", "]},]},
            
            'plot':{"$cond": [{"$eq": ["$plot", ""]},
                                    "$nonExistinField",
                                    "$plot"]},
            
            'poster': {"$cond": [{"$eq": ["$poster", ""]},
                                    "$nonExistinField",
                                    "$poster"]},
            
            'rated':{"$cond": [{"$eq": ["$rating", ""]},
                                    "$nonExistinField",
                                    "$rating"]},
            
            'released':{'$concat': ['datetime.datetime(',
                                    {'$arrayElemAt':[{'$split':['$released','-']},0]},
                                    ', ',
                                    {'$arrayElemAt':[{'$split':['$released','-']},1]},
                                    ', ',
                                    {'$arrayElemAt':[{'$split':['$released','-']},2]},
                                    ", 0, 0)"]},
            
            'runtime': {"$cond": [{'$eq': ["$runtime", ""]}, 0 ,
                                  {'$arrayElemAt':[{'$split': [ "$runtime", " min" ]}, 0]}]},
            
            'title':{"$cond": [{"$eq": ["$title", ""]},
                                    "$nonExistinField",
                                    "$title"]},
            'tomatoes': {
                'lastupdated': "$lastupdated",
                'rating': {'viewer':1, 'rating': '$rating','numReviews':""}
                },
            'type':{"$cond": [{"$eq": ["$type", ""]},
                                    "$nonExistinField",
                                    "$type"]},
            'writers':  {"$cond": [{"$eq": ["$writer", ""]},
                                              "$nonExistinField",
                                              {'$split': ["$writer", ", "]},]},
            'year':{"$cond": [{"$eq": ["$year", ""]},
                                    "$nonExistinField",
                                    "$year"]}    
        }
    }
    ,
    {
        "$match":{"imdb":{'$in':[dict(sorted(imdb_list[x].items(),key=lambda x: x[0])) for x in range(len(imdb_list))]}}
    },
    {
        ### naming the new collection as output ### 
        '$out': "clean_movies_firdaus"
    }
    
]

clear_output()
pprint.pprint(list(col.aggregate(pipeline)))

[]


Beberapa hal yang dilakukan pada projection ini, diantaranya adalah:
- **Menyamakan key** movies_initial dari movies collection
- **Mengurutkan key** movies_initial dari movies collection
- Terdapat beberapa **values yang berbeda** sehingga perlu merubah string menjadi array menggunakan **split** (hanya values yang berbeda)
- Pada kolom awards **memilih array** menggunakan **arrayElemAt** (Memilih array berdasarkan indeksnya) untuk memilih nominations, text, wins
- Melakukan **embedded value** pada **imdb**, dan **tomatoes**. imdb mengembed id, rating, dan value. tomatoes mengembed lastupdated, dan rating
- Melakukukan **condition** pada semua key, dimana conditionnya ialah jika **nilainya None** maka hasilnya **tidak akan keluar**
- Melakukan **matching** dari hasil Key Equalizer untuk menyesuaikan movie_initial dengan movie
- Melakukan **aggregasi** terhadap hasil matching dan projection sehingga hasil dari matching dan projection movie_initial dan movie dimasukkan ke hasil collection movie_clean_firdaus

In [34]:
pprint.pprint(dbs.clean_movies_firdaus.find_one({}))

{'_id': ObjectId('5db0085107645f129de5458b'),
 'awards': {'nominations': None, 'text': '1 win.', 'wins': '1'},
 'cast': 'Charles Kayser, John Ott',
 'countries': ['USA'],
 'directors': ['William K.L. Dickson'],
 'fullPlot': 'A stationary camera looks at a large anvil with a blacksmith '
             'behind it and one on either side. The smith in the middle draws '
             'a heated metal rod from the fire, places it on the anvil, and '
             'all three begin a rhythmic hammering. After several blows, the '
             'metal goes back in the fire. One smith pulls out a bottle of '
             'beer, and they each take a swig. Then, out comes the glowing '
             'metal and the hammering resumes.',
 'genres': ['Short'],
 'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
 'lastupdated': '2015-08-26 00:03:50.133000000',
 'num_mflix_comments': None,
 'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
 'rated': 'UNRATED',
 'released': 'datetime.dat

# 4. Validasi

## 4.1 Semua document pada clean_movies dan movie sama

In [19]:
# Melakukan pengecekan apakah setiap document pada collection Clean_movies_firli ada di collection 
# Dengan melakukan pengecekan setiap document untuk masing-masing collection apakah sama atau tidak
db_clean=dbs.clean_movies_firdaus.find({}).sort('title')
data_movies = db_movies.movies.find({}).sort('title')
if db_clean==data_movies:
    print('Semua document pada kedua collection sama')
else:
    print('Document pada kedua collection ada yang berbeda')

Document pada kedua collection ada yang berbeda


Document berbeda dikarenakan terdapat 2 perbedaan document dari imdb yang dimana adanya document yang **imdbnya di movie ada**, namun di **movies_initial tidak ada**.<br><br>
Mengapa bisa berbeda? akan dibahas pada nomor berikutnya

## 4.2 Banyak document pada clean_movies dan movie sama

In [20]:
data_baru = dbs.clean_movies_firdaus.count()
data_movies = db_movies.movies.count()

print('Jumlah banyaknya dokumen movies: ', data_movies)
print('Jumlah banyaknya dokumen clean_movies_firdaus: ', data_baru)

Jumlah banyaknya dokumen movies:  23541
Jumlah banyaknya dokumen clean_movies_firdaus:  23539


Pada hasil diatas merupakan hasil dari **perbandingan banyaknya jumlah document** antara clean_movies_firdaus dengan movie. Dimana clean_movies_firdaus berasal dari movies_initial yang disamakan dengan movies menggunakan projection.<br><br>
Terdapat perbedaan 2 document dimana clean_movies_firdaus sebanyak **23541** sedangkan movies sebanyak **23541**. Hal itu dikarenakan adanya document yang **imdbnya di movies ada**, tetapi di **movies_initial tidak ada**.

#### Mengecek mengapa terjadi perbedaan dokumen

In [40]:
grup_firdaus = list(dbs['clean_movies_firdaus'].aggregate([{'$sortByCount':'$imdb'}]))
grup_movies = list(moviess['movies'].aggregate([{"$project":
                                                 {'imdb':{"id":'$imdb.id','rating':'$imdb.rating','votes':'$imdb.votes'}}
                                                },{'$sortByCount':'$imdb'}]))

for i in grup_movies:
    if i not in grup_firdaus:
        print(i)

In [39]:
i

{'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189}}

Ternyata, perbedaaan terjadi pada id 5

In [45]:
# collection clean_movies_firdaus
list(dbs.clean_movies_firdaus.find({'imdb':{'id': 5, 'rating': 6.2, 'votes': 1189}}))

[{'_id': ObjectId('5db0085107645f129de5458b'),
  'awards': {'nominations': None, 'text': '1 win.', 'wins': '1'},
  'cast': 'Charles Kayser, John Ott',
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullPlot': 'A stationary camera looks at a large anvil with a blacksmith behind it and one on either side. The smith in the middle draws a heated metal rod from the fire, places it on the anvil, and all three begin a rhythmic hammering. After several blows, the metal goes back in the fire. One smith pulls out a bottle of beer, and they each take a swig. Then, out comes the glowing metal and the hammering resumes.',
  'genres': ['Short'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'num_mflix_comments': None,
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'rated': 'UNRATED',
  'released': 'datetime.datetime(1893, 05, 09, 0, 0)',
  'runtime': '1',
  'title': 'Blacksmith Scene',
  'tom

In [48]:
# collection movies
list(db_movies.movies.find({'imdb':{'id': 5, 'rating': 6.2, 'votes': 1189}}))

[{'_id': ObjectId('5dafd0d3d75e5fa8b23dae21'),
  'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
  'cast': ['Charles Kayser', 'John Ott'],
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith behind it and one on either side. The smith in the middle draws a heated metal rod from the fire, places it on the anvil, and all three begin a rhythmic hammering. After several blows, the metal goes back in the fire. One smith pulls out a bottle of beer, and they each take a swig. Then, out comes the glowing metal and the hammering resumes.',
  'genres': ['Short', 'Long'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'num_mflix_comments': 1,
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'rated': 'UNRATED',
  'released': datetime.datetime(1893, 5, 9, 0, 0),
  'runtime': 1,
  'title': 'Blacksmith Scene II',
  'to

Pada hasil perbandingan diatas antara collection clean_movies_firdaus dengan movies memiliki judul yang sama dan kharakteristik field serta row yang hampir sama. Terdapat perbedaan diantaranya:
- Genre: memiliki key yang sama namun value yang berbeda, pada clean_movies_firdaus ['short'], namun pada movies  ['short',long]
- id: terjadi duplikasi id tetapi dengan object id yang berbeda <br><br>
Mungkin saja, hal itu dikarenakan salah input data, tetapi selain id 5 ini semua datanya sama dengan collection movies

## 4.3 Semua fields pada clean_movies ada pada movie

In [35]:
# how to show all keys MOVIES in one vertical list
key_clean = [list(doc.keys()) for doc in dbs.clean_movies_firdaus.find({})]
key_clean = set([y for z in key_clean for y in z])
key_clean

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullPlot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

In [26]:
# how to show all keys MOVIES in one vertical list
key_movies = [list(doc.keys()) for doc in moviess.find({})]
key_movies =set([j for i in key_movies for j in i])
key_movies

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

In [51]:
print('Banyaknya field pada collection clean_movies_firdaus: ', (len(key_clean)))
print('Banyaknya field pada collection movies: ', (len(key_movies)))

Banyaknya field pada collection clean_movies_firdaus:  22
Banyaknya field pada collection movies:  22


Pada hasil diatas, menunjukkan antara fields clean_movies dan movies. Sehingga hasil fields antara kedua collection tersebut memiliki fields yang sama dengan jumlah 22 dan urutan yang sama

## 4.4 Semua value pada clean_movies sama dengan semua value pada movies dengan urutan yang sama

In [57]:
pprint.pprint(list(moviess.find({'title':'The Cheat'})))

[{'_id': ObjectId('573a1390f29313caabcd5501'),
  'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
  'cast': ['Fannie Ward', 'Sessue Hayakawa', 'Jack Dean', 'James Neill'],
  'countries': ['USA'],
  'directors': ['Cecil B. DeMille'],
  'fullplot': 'Edith Hardy uses charity funds for Wall Street investments in '
              'hopes of buying some new gowns. She loses all the money and '
              'borrows from wealthy oriental Tori. When her husband gives her '
              "the amount she borrowed, Tori won't take it back, branding her "
              'shoulder with a Japanese sign of his ownership. She shoots him. '
              'Her husband takes the blame. In court Edith reveals all to an '
              'angry mob.',
  'genres': ['Drama'],
  'imdb': {'id': 5078, 'rating': 6.5, 'votes': 1660},
  'languages': ['English'],
  'lastupdated': '2015-08-31 00:41:20.670000000',
  'plot': "A venal, spoiled stockbroker's wife impulsively embezzles $10,000 "
          'from the

In [58]:
pprint.pprint(list(dbs.clean_movies_firdaus.find({'title':'The Cheat'})))

[{'_id': ObjectId('5db0085107645f129de545f9'),
  'awards': {'nominations': None, 'text': '1 win.', 'wins': '1'},
  'cast': 'Fannie Ward, Sessue Hayakawa, Jack Dean, James Neill',
  'countries': ['USA'],
  'directors': ['Cecil B. DeMille'],
  'fullPlot': 'Edith Hardy uses charity funds for Wall Street investments in '
              'hopes of buying some new gowns. She loses all the money and '
              'borrows from wealthy oriental Tori. When her husband gives her '
              "the amount she borrowed, Tori won't take it back, branding her "
              'shoulder with a Japanese sign of his ownership. She shoots him. '
              'Her husband takes the blame. In court Edith reveals all to an '
              'angry mob.',
  'genres': ['Drama'],
  'imdb': {'id': 5078, 'rating': 6.5, 'votes': 1660},
  'languages': ['English'],
  'lastupdated': '2015-08-31 00:41:20.670000000',
  'num_mflix_comments': None,
  'plot': "A venal, spoiled stockbroker's wife impulsively embezzles $1

Hasil perbandingan diatas, maka:
- Urutan fields nya sudah sama
- Isi document juga sudah sama

# <center>-TERIMA KASIH-<center>